pkg/build/controller/build_controller.go

```diff
@@ -130,6 +132,16 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 		return err
 	}
 
+	// Handle processing ValueFrom in build strategy environment variables
+	originalEnvVars := buildutil.GetEnvFromStrategy(build)
+	processedEnvVars, err := common.ProcessValueFrom(build, bc.KubeClient)
+	if err != nil {
+		build.Status.Phase = buildapi.BuildPhaseFailed
+		build.Status.Reason = buildapi.StatusReasonInvalidEnvironmentVariable
+		return err
+	}
+	*originalEnvVars = *processedEnvVars
```

Comment:
I think this code is mis-located.  I think it is, or almost certainly will become, a problem to modify the original Build object in memory.  You will need to locate this code at least after the kapi.Scheme.Copy in nextBuildPhase() and do your work on the buildCopy object.